core/trino-main/src/main/java/io/trino/operator/aggregation/AggregationLoopBuilder.java

```diff
@@ -0,0 +1,331 @@
+import java.lang.reflect.Method;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.cycle;
+import static com.google.common.collect.Iterables.limit;
+import static io.airlift.bytecode.Access.FINAL;
+import static io.airlift.bytecode.Access.PRIVATE;
+import static io.airlift.bytecode.Access.PUBLIC;
+import static io.airlift.bytecode.Access.STATIC;
+import static io.airlift.bytecode.Access.a;
+import static io.airlift.bytecode.Parameter.arg;
+import static io.airlift.bytecode.ParameterizedType.type;
+import static io.airlift.bytecode.expression.BytecodeExpressions.constantInt;
+import static io.airlift.bytecode.expression.BytecodeExpressions.constantString;
+import static io.airlift.bytecode.expression.BytecodeExpressions.invokeStatic;
+import static io.airlift.bytecode.expression.BytecodeExpressions.lessThan;
+import static io.airlift.bytecode.expression.BytecodeExpressions.newInstance;
+import static io.trino.sql.gen.BytecodeUtils.invoke;
+import static io.trino.util.CompilerUtils.defineClass;
+import static io.trino.util.CompilerUtils.makeClassName;
+import static java.lang.invoke.MethodHandles.lookup;
+import static java.lang.invoke.MethodType.methodType;
+
+final class AggregationLoopBuilder
+{
+    private AggregationLoopBuilder() {}
+
+    /**
+     * Build a loop over the aggregation function.  Internally, there are multiple loops generated that are specialized for
+     * RLE, Dictionary, and basic blocks, and for masked or unmasked input.  The method handle is expected to have a {@link Block} and int
+     * position argument for each parameter.  The returned method handle signature, will start with as @link {@link AggregationMask}
+     * and then a single {@link Block} for each parameter.
+     */
+    public static MethodHandle buildLoop(MethodHandle function, int stateCount, int parameterCount, boolean grouped)
+    {
+        verifyFunctionSignature(function, stateCount, parameterCount);
+        CallSiteBinder binder = new CallSiteBinder();
+        ClassDefinition definition = new ClassDefinition(
+                a(PUBLIC, STATIC, FINAL),
+                makeClassName("AggregationLoop"),
+                type(Object.class));
+
+        definition.declareDefaultConstructor(a(PRIVATE));
+
+        buildSpecializedLoop(binder, definition, function, stateCount, parameterCount, grouped);
+
+        Class<?> clazz = defineClass(definition, Object.class, binder.getBindings(), AggregationLoopBuilder.class.getClassLoader());
+
+        // it is simpler to find the method with reflection than using lookup().findStatic because of the complex signature
+        Method invokeMethod = Arrays.stream(clazz.getMethods())
+                .filter(method -> method.getName().equals("invoke"))
+                .findFirst()
```

Comment:
Probably better to use `.collect(onlyElement())` from `MoreCollectors` since that will fail if multiple exist